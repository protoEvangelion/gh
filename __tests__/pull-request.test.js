/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: CONTRIBUTORS)
 * SPDX-License-Identifier: BSD-3-Clause
 */

'use strict'

const rewire = require('rewire')
const assert = require('assert')
const pullRequestsUnstable = require('./fixture/pull-request-unstable.json')
const pullRequestsInfo = require('./fixture/pull-request-info.json')
const { runCmd } = require('./testUtils')

describe('E2E: Pull Request Module Test', () => {
    it('List PRs `gh pr`', done => {
        expect(runCmd('bin/gh.js pr')).toMatchSnapshot()
        done()
    })

    it('List PRs `gh pr --detailed`', done => {
        expect(runCmd('bin/gh.js pr --detailed')).toMatchSnapshot()
        done()
    })

    it('Get PR Info `gh pr --info 50`', done => {
        expect(runCmd('bin/gh.js pr --info 50')).toMatchSnapshot()
        done()
    })

    it('Comment on PR `gh pr 50 --comment "Just started reviewing :)"`', done => {
        expect(runCmd('bin/gh.js pr 50 --comment "Just started reviewing :)"')).toMatchSnapshot()
        done()
    })

    it('Open PR `gh pr 50 --open`', done => {
        expect(runCmd('bin/gh.js pr 50 --open')).toMatchSnapshot()
        done()
    })

    it('Close PR `gh pr 50 --close`', done => {
        expect(runCmd('bin/gh.js pr 50 --close')).toMatchSnapshot()
        done()
    })

    it('Submit PR `gh pr -s protoEvangelion -b master -t "pr title" -D "pr description"`', done => {
        expect(
            runCmd('bin/gh.js pr -s protoEvangelion -b master -t "pr title" -D "pr description"')
        ).toMatchSnapshot()
        done()
    })
})

describe('Pull Requests Module Tests', function() {
    const pullRequest = rewire('../lib/cmds/pull-request')

    it('should fail to get pull request', function() {
        var pr = new pullRequest.Impl({
            repo: 'senna.js',
        })

        pullRequest.__with__({
            'logger.log': function() {
                assert.fail('Expected test to fail.')
            },
            'logger.warn': function() {},
            base: {
                github: {
                    pullRequests: {
                        get: function(payload, callback) {
                            callback(new Error('Should fail to get pull request.'), undefined)
                        },
                    },
                },
            },
        })(function() {
            pr.get('liferay', 'senna.js', '36')
        })
    })

    it('should get pull request with mergeable state clean', function() {
        var cleanLogFound = false,
            pr = new pullRequest.Impl({
                repo: 'senna.js',
            })

        pr.options.info = true

        pullRequest.__with__({
            'logger.log': function() {
                // only evaluate that the clean has a green log message
                if (arguments[0].indexOf('clean') > 1) {
                    cleanLogFound = true

                    assert.strictEqual(arguments[0], 'Mergeable (clean)')
                }
            },
            'logger.warn': function() {
                assert.fail('Expected test to pass.')
            },
            base: {
                github: {
                    pullRequests: {
                        get: function(payload, callback) {
                            callback(undefined, pullRequestsInfo)
                        },
                    },
                },
            },
        })(function() {
            pr.get('liferay', 'senna.js', '36')
        })

        if (cleanLogFound === false) {
            assert.fail('Clean log was not found.')
        }
    })

    it('should get pull request with mergeable state unstable', function() {
        var pr = new pullRequest.Impl({
            repo: 'senna.js',
        })

        pr.options.info = true

        pullRequest.__with__({
            'logger.log': function() {},
            'logger.warn': function() {
                assert.strictEqual(arguments[0], 'Not mergeable (unstable)')
            },
            base: {
                github: {
                    pullRequests: {
                        get: function(payload, callback) {
                            callback(undefined, pullRequestsUnstable)
                        },
                    },
                },
            },
        })(function() {
            pr.get('liferay', 'senna.js', '78')
        })
    })

    it('should only get the issue number from a properly prefixed branch', function() {
        var pr = new pullRequest.Impl({ repo: 'senna.js' })

        assert.equal(pr.getPullRequestNumberFromBranch_('pr-12345', 'pr-'), '12345')
        assert.equal(pr.getPullRequestNumberFromBranch_('abcpr-12345', 'pr-'), undefined)
    })
})
