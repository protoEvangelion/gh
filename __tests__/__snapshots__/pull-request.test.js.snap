// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E: Pull Request Module Test Close PR \`gh pr 50 --close\` 1`] = `
"Closing pull request #50
https://github.com/protoEvangelion/gh/pull/50
"
`;

exports[`E2E: Pull Request Module Test Comment on PR \`gh pr 50 --comment "Just started reviewing :)"\` 1`] = `
"Adding comment on pull request #50
https://github.com/protoEvangelion/gh/pull/50#issuecomment-000000001
"
`;

exports[`E2E: Pull Request Module Test Get PR Info \`gh pr --info 50\` 1`] = `
"#50 pr2 @protoEvangelion (a year ago)
"
`;

exports[`E2E: Pull Request Module Test List PRs \`gh pr --detailed\` 1`] = `
"Listing open pull requests on protoEvangelion/gh
protoEvangelion/gh
master (1)
┌─────────┬────────────────────────────────────────┬─────────────────────┬───────────────┬────────┐
│    #    │ Details                                │ Author              │ Opened        │ Status │
├─────────┼────────────────────────────────────────┼─────────────────────┼───────────────┼────────┤
│   #55   │ Title:                                 │ @protoEvangelion    │ a year ago    │   ✓    │
│         │                                        │                     │               │        │
│         │ pr title                               │                     │               │        │
│         │                                        │                     │               │        │
│         │ Body:                                  │                     │               │        │
│         │                                        │                     │               │        │
│         │ pr description                         │                     │               │        │
│         │                                        │                     │               │        │
│         │ https://github.com/protoEvangelion/gh… │                     │               │        │
└─────────┴────────────────────────────────────────┴─────────────────────┴───────────────┴────────┘
"
`;

exports[`E2E: Pull Request Module Test List PRs \`gh pr\` 1`] = `
"Listing open pull requests on protoEvangelion/gh
protoEvangelion/gh
master (1)
┌─────────┬────────────────────────────────────────┬─────────────────────┬───────────────┬────────┐
│    #    │ Title                                  │ Author              │ Opened        │ Status │
├─────────┼────────────────────────────────────────┼─────────────────────┼───────────────┼────────┤
│   #55   │ pr title                               │ @protoEvangelion    │ a year ago    │   ✓    │
└─────────┴────────────────────────────────────────┴─────────────────────┴───────────────┴────────┘
"
`;

exports[`E2E: Pull Request Module Test Open PR \`gh pr 50 --open\` 1`] = `
"Opening pull request #50
https://github.com/protoEvangelion/gh/pull/50
"
`;

exports[`E2E: Pull Request Module Test Submit PR \`gh pr -s protoEvangelion -b master -t "pr title" -D "pr description"\` 1`] = `
"Submitting pull request to @protoEvangelion
https://github.com/protoEvangelion/gh/pull/55
[hook] gh pr --browser  --user protoEvangelion --repo gh --number 55
"
`;
